Add some improvements

mmattel opened the pull request:
This PR adds some improvements:

* Add `exclude.txt` and `shrink-backup.log` to .gitignore
This allows to use the repo as base
* Rename `exclude.txt` --> `exclude.txt.example`
This can now be used as base for your own file without triggering a git change
* Add a README change about that fact
* Fix the `--timeout=10` to `--timeout=120`

Note that increasing the timeout was necessary because I run the script to a target via NFS over WIFI.

It might be an idea to make this value configurable via a CLI option.

Note that the following directories could be also added to the exclude list. I did in my env but not added this to the example list. My .cache dirs have ~300M which is not that much but at least something...
```
home/**/.cache/*
root/.cache/*
```

UnconnectedBedna commented:
Honestly, I was going to remove .gitignore completely since it makes no sense using it (and is incorrect, exclude.txtis still tracked in git so it wont matter anyway, I would need to untrack it on both branches so your changes would not do anything anyway), but here are my thoughts of your suggested changes.

- shrink-backup.log is not tracked by git so it makes no sense/no use having it in .gitignore
- exclude.txt would not get updated with the upcoming changes in the next release. exclude.txt has never been changed before, but when it does it should absolutely get pushed with an update. See: https://github.com/UnconnectedBedna/shrink-backup/commits/main/exclude.txt
- Forcing the user to both edit and change name of exclude.txt is not a good idea, in that case it might as well be deleted completely and user would have to create the file from documentation.
- a 2 min timeout is way too long. If half a minute passes (the timeout is 30s, not 10, it's only in the debug line it is incorrect) without rsync being able to function is more than enough.

If a user makes changes to exclude.txt, the only problem will be if the file is changing on the repo (witch it will in the next release for the first time ever) and that will be informed about in the patch notes of how to handle. It will also become an issue if switching branch to testing, but a user doing that is expected to understand the implications of switching and using an unstable branch.

Wifi should not force you to have a 2 min timeout, if that happens, there is something else going on and should not be fixed this way on the default repo. If a user has that slow of a connection, they can manually make changes to the script like you did.

You did however point out the timer for rsync in the debug line was incorrect at 10s.

All of this would be better to put in the wiki (witch is open for anyone to edit)

mmattel commented:
Just to let you know, I iterated the timeout to 60, 90 and 120.
With 120 it worked which took a lot of testing time because I needed to try a complete the backup.

When using git as base without installing (copying files), you must treat it as this PR proposes.
I removed `exclude.txt` from tracking and added `exclude.txt.example` instead including an entry in .gitignore.
You will find many repos who exactly do the same. Exclude a particular config file but provide a tracked example.

You are right, the changes would need to go into the other branches too.

Fixed the missing slash in the exclude list and also updated the .cache items in the main post

UnconnectedBedna commented:
> With 120 it worked which took a lot of testing time because I needed to try a complete the backup.

I understand and appreciate your dedication and testing. :+1: 
But you are the exception, not the rule in this situation. The year is 2025... xD

> Making the timeout a config value was an idea I posted.

That actually crossed my mind too, but again, this is a very exceptional situation, it would be better information in documentation (wiki) than an option.

> When using git as base without installing (copying files), you must treat it as this PR proposes.
> I removed exclude.txt from tracking and added exclude.txt.example instead including an entry in .gitignore.

I think you misunderstand how that works. If the name of the file were to be changed on the repo (ie rename it from exclude.txt > exclude.txt.example, it would not be necessary to be in the .gitignore either, for the same reason as the log-file. exclude.txt would no longer be tracked since it does not exist on the github repo.

So unless you have better arguments for any of these changes, this pr will be closed.

**But I REALLY appreciate your willingness to contribute!**

UnconnectedBedna commented:
You know what, since you are a perfect person to test this out, I can give you an idea about something that might be usefull as an option in the script.

Read this: https://www.raysync.io/news/rsync-is-slow/

Suggestion to test:
> Disable compression with the --no-compress option when transferring over high-speed networks.
&
> Use the --whole-file option to bypass delta calculations for small files.

The first option is mostly if your system bottlenecks because your docker container in combination with the transfer because of cpu usage, check with `top` or `htop` to see how your systems act while transferring files with rsync.

The second option could actually be useful when having docker containers due to the massive amount of tiny files.

mmattel commented:
I tested
`-W, --whole-file            copy files whole (w/o delta-xfer algorithm)`
as alternative to the timeout but without success. Setting the timeout was key.

mmattel commented:
> But you are the exception, not the rule in this situation. The year is 2025... xD

I am a professional doc writer and need by "design" deep dive into stuff. Finding things others dont and I am "magic" to the "this is the execptional thing" stuff...

mmattel commented:
> And if there are no changes made to the file on the repo when doing a git pull there is no issue having changed the file locally. I think that is where the confusion lies.

I have changed content in the exclude.txt file which extends the one from origin, therefore it is different and therefore it is tracked if not .gitignored... But lets agree on disagree :sweat_smile: 

Note that after rebooting the RPi and running sb again including your changes in testing and my additions from this PR, all is doing fine :smile:

mmattel commented:
> You can again test that if you don't believe me by making sure you are up to date with a git reset --hard, edit exclude.txt and do one more git pull, it will work without issues.

This is not the point. If I would do some contributions, all changed files would be highlighted including log and exclude. To avoid any problems such when using `git add .`, I suggest using the proposed method...

UnconnectedBedna commented:
Oh ****, I'm sorry, I didn't mean to edit YOUR comment! I thought it was mine. xD
If you use `git add` you are AWARE you are not up to par with the cloned repo, your argument is mute.
Just don't git add, there is no reason for a normal user to do that!

You are making a chicken out of a feather.

UnconnectedBedna commented:
I made tests on a test repo and learned something today. (I am far from a git expert so this is not very surprising)
The .gitignore is mainly for local use, ie changes made to files in .gitignore are ignored when pushing, not the other way around.
And if a file exists on a repo that is cloned, it will be tracked and .gitignore will have no effect unless making the file untracked locally witch then makes other problems arisie.

As the primary focus is to keep users up to date with all files .gitignore will be removed from the repo and the user will have to implement it themselves if they need it.

### However
I have been thinking about your suggestion about renaming exclude.txt to exclude.txt.example and have come to the conclusion that maybe that idea is not so bad after all. But I do think the filename exclude_example.txt is better so the file still have the .txt suffix for easy editing, ie users can use a file manager and double click it in a gui environment and it will open in a text editor.
So if you want to revert your changes into that I will accept the pr.

So:
- Remove .gitignore since it will no longer be needed
- Change name of exclude.txt to exclude_example.txt
- Edit README.md to reflect information about renaming exclude_example.txt under [-t (exclude.txt)](https://github.com/UnconnectedBedna/shrink-backup?tab=readme-ov-file#-t-excludetxt)
<hr>

Then if you also want to create an option of user inputting rsync line manually (for example --rsync) I can work with you to make that happen, but should be done in a different pr.
The best way is probably to ask for user to input rsync line after starting the script with --rsync because example lines can be echoed that the user can copy > paste > edit easily

UnconnectedBedna commented:
> Let me give some insights. .gitignore defines all files and folders that can be added/removed/changed without getting tracked - independent if they are upstream (your repo) or downstream (the local clone). As .gitignore is also a trackable file like any others, the same rules apply. The name itself tells git about its use. Now, if you remove this file upstream - which must be done there, all files in upstream are tracked and synchronized with downstream. If upstream does not contain a .gitignore and a user adds .gitignore downstream, it gets tracked downstream. The only thing one can do downstream is to create a local .gitignore and include .gitignore in .gitignore which is possible. Allow me some advise: this is a VERY BAD idea and should not be done. There are good reasons for heaving it up- and downstream which why it was added to git. The impact of removing .gitignore is far worse than managing it, which is quite easy. Do you really want to go the path removing it?

Really?
Then why are the changes made upstream to exclude.txt a couple of days ago being pulled in a `git pull`?

I have proven that it only becomes an issue if files not existing upstream are manually added by add/commit downstream, and in that case, it's a conscious choice by the user to step out of tree an not the repos responsibility.

```
$ git checkout testing
Switched to branch 'testing'
Your branch is up to date with 'origin/testing'.

# Edited exclude.txt

$ cat exclude.txt
/lost+found
/proc/*
/sys/*
/dev/*
/tmp/*
/run/*
/mnt/*
/media/*
/var/swap
/snap/*
/test


$ git status
On branch testing
Your branch is up to date with 'origin/testing'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   exclude.txt

no changes added to commit (use "git add" and/or "git commit -a")


$ git pull
remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (5/5), done.
remote: Compressing objects: 100% (1/1), done.
remote: Total 3 (delta 2), reused 3 (delta 2), pack-reused 0 (from 0)
Unpacking objects: 100% (3/3), 1.14 KiB | 1.14 MiB/s, done.
From https://github.com/UnconnectedBedna/shrink-backup
   f788cb4..c7bc1ae  testing    -> origin/testing
Updating f788cb4..c7bc1ae
Fast-forward
 shrink-backup | 35 +++++++++++++++++++++--------------
 1 file changed, 21 insertions(+), 14 deletions(-)
```

See: https://git-scm.com/docs/gitignore

> **NAME**
> 
> gitignore - Specifies intentionally untracked files to ignore
> 
> **DESCRIPTION**
> 
> A gitignore file specifies intentionally untracked files that Git should ignore. Files already tracked by Git are not affected; see the NOTES below for details.
> 
> **NOTES**
> 
> The purpose of gitignore files is to ensure that certain files not tracked by Git remain untracked.
> 
> To stop tracking a file that is currently tracked, use git rm --cached to remove the file from the index. The filename can then be added to the .gitignore file to stop the file from being reintroduced in later commits.

.gitignore is there for development.

To get around the problem a user with an edited exclude.txt might face when switching branch I accepted the name-change. But this also means that normal users won't automatically get updates to their file that they use (get notified when using `git pull` that changes are made), it's not a perfect solution but it's a compromise I'm willing to make.
Maybe a `diff exclude_example.txt exclude.txt` could be invoked every time the script runs (if exclude.txt exists) to make users aware of manual changes they have made and new additions to exclude_example.txt to get around that. I have a feeling 90% of users won't read patch-notes.. xD

And for contributors/developers, just don't use `git add .` but rather add the specific file/s manually, for example `git add shrink-backup` if changes are made to the script. We are not talking about a massive repo with hundreds of files here, shrink backup has 5 files. (6 with .gitignore)

TBH, I very rarely `git push`, I just change the files directly on github, (unless it involves chmod changes and reverts) because of the tiny amount of files. The diff functionality and README.md preview is actually pretty nice using a web browser.

mmattel commented:
I'll do the changes required in the next days (was on sickleave)

UnconnectedBedna commented:
If you want to make these changes to be included in the upcoming release, you should probably get started with them, I am getting close to the next release.

As I said earlier, I am willing to help out with the curl install script if you find it confusing.

The `--rsync` option just got implemented on the testing branch so it should be easy for you to change the timeout now. :)

UnconnectedBedna commented:
> I am missing a discription in the repos readme

There are more than just this missing in the readme, I haven't updated it yet.

> For testing this is ok, but when the parameters are set, as admin, I just want to handover the new options for rsync defined. This is needed for automated processing. How to do that? Maybe it needs a rsync-print (--rprint) parameter as you implemented but the real values are handed over by --rsync. Just dumping my ideas...

Not a bad idea, but becomes complicated when you start looking into the code. (unless a tag that modifies the actual code in the script, see below)
I feel like a sysadmin that want that change the default value that way will edit the script and just change whatever they want hard-coded manually.
But feel free to come with solutions in the code.

The whole rsync functionality has been revamped into variable that are then ran with eval and can be found here:
(`$RSYNC_CUSTOM_LINE` is just used for the user to be able to see the line again before pressing "y" and start the backup)
https://github.com/UnconnectedBedna/shrink-backup/blob/93ad78a3f07417caf9bca1e723888551884799f2/shrink-backup#L2567-L2607
And:
https://github.com/UnconnectedBedna/shrink-backup/blob/93ad78a3f07417caf9bca1e723888551884799f2/shrink-backup#L1410-L1445

UnconnectedBedna commented:
v1.3 released
